### PR TITLE
feat: Query Market Size

### PR DESCRIPTION
**Pull request by RafilxTenfen**
## Description

- Add structures to retrieve the Market Size of a specific denom

ref: #343 

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/umee/leverage/v1beta1/query.proto [hunk 1]
```diff
@@ -37,6 +37,11 @@ service Query {
     option (google.api.http).get = "/umee/leverage/v1beta1/lend_apy";
   }
 
+  // MarketSize queries for the Market Size of a specified denomination.
```
Comment (toteki) [inline proto/umee/leverage/v1beta1/query.proto:40]:
```suggestion
  // MarketSize queries for the Market Size in USD of a specified denomination.
```

File: proto/umee/leverage/v1beta1/query.proto [hunk 2]
```diff
@@ -102,6 +107,19 @@ message QueryLendAPYResponse {
   [(gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec", (gogoproto.nullable) = false];
 }
 
+// QueryMarketSizeRequest defines the request structure for the Market Size
```
Comment (toteki) [inline proto/umee/leverage/v1beta1/query.proto:110]:
```suggestion
// QueryMarketSizeRequest defines the request structure for the MarketSize
```

File: proto/umee/leverage/v1beta1/query.proto [hunk 3]
```diff
@@ -102,6 +107,19 @@ message QueryLendAPYResponse {
...
+// gRPC service handler.
+message QueryMarketSizeRequest {
+  string denom = 1;
+}
+
+// QueryMarketSizeResponse defines the response structure for the Market Size in USD
```
Comment (toteki) [inline proto/umee/leverage/v1beta1/query.proto:116]:
```suggestion
// QueryMarketSizeResponse defines the response structure for the MarketSize
```

File: x/leverage/client/tests/suite.go [hunk 4]
```diff
@@ -712,6 +712,42 @@ func (s *IntegrationTestSuite) TestQueryBorrowAPY() {
 	runTestQueries(s, testCasesBorrowAPY)
 }
 
+func (s *IntegrationTestSuite) TestQueryMarketSize() {
+	testCasesMarketSize := []testQuery{
```
Comment (toteki) [inline x/leverage/client/tests/suite.go:716]: This test can be improved by using `setup` commands with a `Lend` (and `cleanup` commands with a `Withdraw`). That way, the market size will be nonzero.

Note that the simulated `Umee` oracle price is 34.21 during CLI tests, so you should come out with an expected value of `sdk.MustNewDecFromStr("34.21")` if you lend `1000000uumee` in the setup.

Comment (RafilxTenfen) [reply x/leverage/client/tests/suite.go]: done at https://github.com/umee-network/umee/pull/376/commits/bcf42b0e98935e745ff0ae40024fd77e4230cee3

Comment (RafilxTenfen) [reply x/leverage/client/tests/suite.go]: add the clean up command in https://github.com/umee-network/umee/pull/376/commits/7a5c33f66a267aeb91567c7035a1e353c0af22d4

File: proto/umee/leverage/v1beta1/query.proto [hunk 5]
```diff
@@ -102,6 +107,19 @@ message QueryLendAPYResponse {
...
+// QueryMarketSizeRequest defines the request structure for the Market Size in USD
+// gRPC service handler.
+message QueryMarketSizeRequest {
+  string denom = 1;
+}
+
+// QueryMarketSizeResponse defines the response structure for the Market Size in USD
+// gRPC service handler.
+message QueryMarketSizeResponse {
+  string marketSizeUSD = 1
```
Comment (alexanderbez) [inline proto/umee/leverage/v1beta1/query.proto:119]: proto fields are snake-cased :)

```suggestion
  string market_size_usd = 1
```

File: x/leverage/client/cli/query.go [hunk 6]
```diff
@@ -340,6 +341,39 @@ func GetCmdQueryBorrowAPY() *cobra.Command {
 	return cmd
 }
 
+// GetCmdQueryMarketSize returns a CLI command handler to query for the
+// Market Size of a specific token.
+func GetCmdQueryMarketSize() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "market-size [denom]",
+		Args:  cobra.ExactArgs(1),
+		Short: "Query for the market size of a specified denomination in USD",
```
Comment (alexanderbez) [inline x/leverage/client/cli/query.go:350]:
```suggestion
		Short: "Query for the USD market size of a specified denomination",
```

File: x/leverage/keeper/grpc_query.go [hunk 7]
```diff
@@ -136,6 +136,35 @@ func (q Querier) LendAPY(
...
+	if req.Denom == "" {
+		return nil, status.Error(codes.InvalidArgument, "invalid denom")
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	if !q.Keeper.IsAcceptedToken(ctx, req.Denom) {
+		return nil, status.Error(codes.InvalidArgument, "not accepted Token denom")
+	}
+
+	uDenom := q.Keeper.FromTokenToUTokenDenom(ctx, req.Denom)
```
Comment (alexanderbez) [inline x/leverage/keeper/grpc_query.go:155]:
```suggestion
	uTokenDenom := q.Keeper.FromTokenToUTokenDenom(ctx, req.Denom)
```

File: x/leverage/keeper/grpc_query.go [hunk 8]
```diff
@@ -136,6 +136,35 @@ func (q Querier) LendAPY(
...
+	marketSizeCoin, err := q.Keeper.ExchangeUToken(ctx, q.Keeper.TotalUTokenSupply(ctx, uDenom))
+	if err != nil {
+		return nil, err
+	}
+
+	marketSizeUSD, err := q.Keeper.TokenValue(ctx, marketSizeCoin)
+	if err != nil {
+		return nil, err
+	}
+	return &types.QueryMarketSizeResponse{MarketSizeUSD: marketSizeUSD}, nil
```
Comment (alexanderbez) [inline x/leverage/keeper/grpc_query.go:165]:
```suggestion

	return &types.QueryMarketSizeResponse{MarketSizeUSD: marketSizeUSD}, nil
```

**Review by alexanderbez — CHANGES_REQUESTED** [review]
Nice work @RafilxTenfen 👏

File: proto/umee/leverage/v1beta1/query.proto (hunk 5 above)
Comment (RafilxTenfen) [reply proto/umee/leverage/v1beta1/query.proto]: this way the attribute stays `MarketSizeUsd` instead of `MarketSizeUSD` is that okay?

**Review by alexanderbez — APPROVED** [review]
🎉

**Comment by alexanderbez** [comment]
I'll merge this once I get #369 taken care of

**Pull request merged**